test/xbgpu/test_engine.py

```diff
@@ -1025,15 +1005,20 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
         # assert_allclose converts to float, which bloats memory usage.
         # To keep it manageable, compare a batch at a time.
         for i in range(beam_results.shape[0]):
-            for j in range(beam_results.shape[1]):
+            # NOTE: As per the explanation at the end of `_send_data`, we
+            # only verify up the the penultimate heap's data for each `beam_result`
+            # because the final heap is sent coincidentally - not because it is
+            # expected to have sane data in it.
+            for j in range(beam_results.shape[1] - 1):
```

Comment:
Again, you're hardcoding the assumption about how must extra is sent. How about just iterating up to the length of `expected_beams`?